parser-typechecker/src/Unison/Codebase/FileCodebase/SlimCopyRegenerateIndex.hs

```diff
@@ -304,16 +308,17 @@ syncToDirectory' getV getA srcPath destPath newRoot =
   tellError :: forall m a. MonadWriter (Set a) m => a -> m ()
   tellError = Writer.tell . Set.singleton
 
--- Use State and Lens to do some specified thing at most once, to create a file.
-ifNeedsSyncing :: forall m s h. (MonadIO m, MonadState s m, Ord h)
-               => h
-               -> CodebasePath
-               -> (CodebasePath -> h -> FilePath) -- done if this filepath exists
-               -> SimpleLens s (Set h) -- lens to track if `h` is already done
-               -> (h -> m ()) -- do!
-               -> m ()        -- don't
-               -> m ()
-ifNeedsSyncing h destPath getFilename l doSync dontSync =
-  ifM (use (l . to (Set.member h))) dontSync $ do
-    l %= Set.insert h
-    ifM (doesFileExist (getFilename destPath h)) dontSync (doSync h)
+  -- Use State and Lens to do some specified thing at most once, to create a file.
+  ifNeedsSyncing :: forall m s h. (MonadIO m, MonadState s m, Ord h)
+                 => h
+                 -> CodebasePath
+                 -> (CodebasePath -> h -> FilePath) -- done if this filepath exists
+                 -> SimpleLens s (Set h) -- lens to track if `h` is already done
+                 -> (h -> m ()) -- do!
+                 -> m ()        -- don't
+                 -> m ()
+  ifNeedsSyncing h destPath getFilename l doSync dontSync =
+    ifM (use (l . to (Set.member h))) dontSync $ do
+      l %= Set.insert h
+      if mode == SyncMode.Complete then doSync h
+      else ifM (doesFileExist (getFilename destPath h)) dontSync (doSync h)
```

Comment:
The `if` on line 323 is the only interesting change here.

Reply:
Do we need to do something special here to make sure we don't visit the same hash over and over again every time it appears somewhere in the namespace tree or the history? The short-circuiting version relies on the existence of the file to prevent this worst case, which can be exponential in the worst case (it's based on the number of unique paths to a node in a DAG), so do we need to do something extra here to prevent that? Like maybe a `seen : Set Hash` that gets used in the `SyncMode.Complete` case? (Or maybe you are already doing this...)

I have a question case

Reply:
Right, on lines 321 and 322 we first check/update a ~`seen : Set h`.  So we only get to line 323 once.  Before this PR, that `Set` was being used to skip even the `doesFileExist` check.  Does that make sense?